### PR TITLE
[Cocoa] [Shape Detection] WebKit crashes on launch when on the base system

### PR DESCRIPTION
**Pull request by litherum**
#### fcedcb75b10f017f98ec88d832d9fb01dd8921b6
<pre>
[Cocoa] [Shape Detection] WebKit crashes on launch when on the base system
<a href="https://bugs.webkit.org/show_bug.cgi?id=256287">https://bugs.webkit.org/show_bug.cgi?id=256287</a>
rdar://108870362

Reviewed by Aditya Keerthi.

Turns out the base system doesn&apos;t have Vision.framework or CoreML.framework. We were crashing
because we were statically linking with CoreML.framework. This patch makes us use soft linking
instead.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm:
(WebCore::ShapeDetection::configureRequestToUseCPUOrGPU):
* Source/WebCore/PAL/pal/cocoa/CoreMLSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/CoreMLSoftLink.mm:
* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.mm:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/263661@main">https://commits.webkit.org/263661@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d20224c9f9295b34b0fbf8ac6f263973a94d0bb4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5387 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5715 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6927 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/5418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5386 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5505 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/6289 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5555 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4824 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6956 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3034 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4831 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/11984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/4467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4909 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6546 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4967 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5341 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4385 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5490 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4802 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1369 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1286 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8902 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5647 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5163 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1487 "Passed tests") |

### CODE REVIEW

**Review by pxlcoder — APPROVED** [review]

File: Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm [hunk 1]
```diff
@@ -79,15 +80,15 @@ void configureRequestToUseCPUOrGPU(VNRequest *request)
     for (VNComputeStage computeStage in supportedComputeStageDevices) {
         bool set = false;
         for (id<MLComputeDeviceProtocol> device in supportedComputeStageDevices[computeStage]) {
-            if ([device isKindOfClass:MLGPUComputeDevice.class]) {
+            if ([device isKindOfClass:[PAL::getMLGPUComputeDeviceClass() class]]) {
```
Comment (pxlcoder) [inline Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: I think the `+class` here is redundant?

Comment (litherum) [reply Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: it isn't, sadly :(

**Comment by litherum** [comment]
I'll update the copyrights before landing.

File: Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm (hunk 1 above)
Comment (pxlcoder) [reply Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: Hmm, I see a bunch of `isKindOfClass:PAL::get` in the code. Should we be concerned...

(and no hits for `isKindOfClass:[PAL::get`)

Comment (litherum) [reply Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: Interesting! Let me try again! I assumed it wasn't redundant, because the previous code said `[MLGPUComputeDevice class]` but maybe the new way just works!

Comment (pxlcoder) [reply Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: Yeah, it should just work.

The `get...Class()` is really just a wrapper (that also caches the return value) for `objc_getClass`, which returns a `Class`. (see `wtf/cocoa/SoftLinking.h`)

The only difference here would be if `MLGPUComputeDevice` were to override `+class`. But I think that would just be asking for trouble 😅

Comment (litherum) [reply Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: It compiles! Let me see if it does what it's supposed to do...

Comment (litherum) [reply Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm]: Yes! It works! \o/
